analyzers/src/SonarAnalyzer.CSharp/Rules/DisposableTypesNeedFinalizers.cs

```diff
@@ -35,28 +36,22 @@ public sealed class DisposableTypesNeedFinalizers : SonarDiagnosticAnalyzer
         private const string DiagnosticId = "S4002";
         private const string MessageFormat = "Implement a finalizer that calls your 'Dispose' method.";
 
-        private static readonly DiagnosticDescriptor Rule
-            = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+        private static readonly ImmutableArray<KnownType> NativeHandles = ImmutableArray.Create(
+            KnownType.System_IntPtr,
+            KnownType.System_UIntPtr,
+            KnownType.System_Runtime_InteropServices_HandleRef);
 
-        private static readonly ImmutableArray<KnownType> NativeHandles =
-            ImmutableArray.Create(KnownType.System_IntPtr,
-                                  KnownType.System_UIntPtr,
-                                  KnownType.System_Runtime_InteropServices_HandleRef);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
         protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(c =>
             {
-                if (c.ContainingSymbol.Kind != SymbolKind.NamedType)
-                {
-                    return;
-                }
-
                 var declaration = (TypeDeclarationSyntax)c.Node;
-                var classSymbol = c.SemanticModel.GetDeclaredSymbol(declaration);
-
-                if (classSymbol.Implements(KnownType.System_IDisposable)
+                if (!c.IsRedundantPositionalRecordContext()
+                    && c.SemanticModel.GetDeclaredSymbol(declaration) is var classSymbol
+                    && classSymbol.Implements(KnownType.System_IDisposable)
```

Comment:
We already do have that symbol as input so we don't need to re-query it
```suggestion
                    && ((ITypeSymbol)c.ContainingSymbol).Implements(KnownType.System_IDisposable)
```